xarray/core/common.py

```diff
@@ -347,6 +347,34 @@ def _ipython_key_completions_(self) -> list[str]:
         return list(items)
 
 
+class TreeAttrAccessMixin(AttrAccessMixin):
+    """Mixin class that allows getting keys with attribute access"""
+
+    # TODO: Ensure ipython tab completion can include both child datatrees and
+    # variables from Dataset objects on relevant nodes.
+
+    __slots__ = ()
+
+    def __init_subclass__(cls, **kwargs):
+        """Verify that all subclasses explicitly define ``__slots__``. If they don't,
+        raise error in the core xarray module and a FutureWarning in third-party
+        extensions.
+        """
+        if not hasattr(object.__new__(cls), "__dict__"):
+            pass
+        # TODO Rework DataTree to avoid __dict__.
+        # elif cls.__module__.startswith("xarray."):
+        #     raise AttributeError(f"{cls.__name__} must explicitly define __slots__")
+        # else:
+        #     cls.__setattr__ = cls._setattr_dict
+        #     warnings.warn(
+        #         f"xarray subclass {cls.__name__} should explicitly define __slots__",
+        #         FutureWarning,
+        #         stacklevel=2,
+        #     )
+        # super().__init_subclass__(**kwargs)
```

Comment:
This comment is already captured in the issue https://github.com/pydata/xarray/issues/9068 right?

Reply:
It is now, yup. I can rework this TODO to point back to the feature instead. That feels like the right thing to do.